src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -259,56 +256,51 @@ internal static ProcessInfo[] GetProcessInfos(Predicate<int>? processIdFilter =
+                        // ArrayPool does not guarantee 64-byte alignment, so we take the address of first 64-byte aligned element
+                        // we round up the address of the pinned array to the nearest multiple of 64
+                        Debug.Assert(bufferSize > 64);
+                        byte* alignedBufferPtr = (byte*)(((nint)bufferPtr + 63) & ~63);
+                        int firstAlignedElementIndex = (int)(alignedBufferPtr - bufferPtr);
+
+                        uint status = Interop.NtDll.NtQuerySystemInformation(
+                            Interop.NtDll.SystemProcessInformation,
+                            alignedBufferPtr,
+                            (uint)(buffer.Length - firstAlignedElementIndex),
+                            &requiredSize);
+
+                        if (status != Interop.NtDll.STATUS_INFO_LENGTH_MISMATCH)
                         {
-                            uint status = Interop.NtDll.NtQuerySystemInformation(
-                                Interop.NtDll.SystemProcessInformation,
-                                bufferPtr,
-                                (uint)(buffer.Length * sizeof(long)),
-                                &requiredSize);
-
-                            if (status != Interop.NtDll.STATUS_INFO_LENGTH_MISMATCH)
+                            // see definition of NT_SUCCESS(Status) in SDK
+                            if ((int)status < 0)
                             {
-                                // see definition of NT_SUCCESS(Status) in SDK
-                                if ((int)status < 0)
-                                {
-                                    throw new InvalidOperationException(SR.CouldntGetProcessInfos, new Win32Exception((int)status));
-                                }
-
-                                // Parse the data block to get process information
-                                processInfos = GetProcessInfos(MemoryMarshal.AsBytes<long>(buffer), processIdFilter);
-                                break;
+                                ArrayPool<byte>.Shared.Return(buffer);
+
+                                throw new InvalidOperationException(SR.CouldntGetProcessInfos, new Win32Exception((int)status));
                             }
+
+                            // Parse the data block to get process information
+                            processInfos = GetProcessInfos(buffer.AsSpan(firstAlignedElementIndex), processIdFilter);
+                            break;
                         }
                     }
-
-                    buffer = null;
-                    bufferSize = GetNewBufferSize(bufferSize, (int)requiredSize);
                 }
-            }
-            finally
-            {
-                // Cache the final buffer for use on the next call.
-                Interlocked.Exchange(ref CachedBuffer, buffer);
+
+                ArrayPool<byte>.Shared.Return(buffer);
```

Comment:
I think this return operation is in the while loop. You're not assigning `buffer` to a new allocation, so if I'm understanding this correctly you'll be reusing the returned instance?

Reply:
great catch, it was a bug in the logic. fixed